R/plotPrevalence.R

```diff
@@ -0,0 +1,322 @@
+                     call. = FALSE)
+            }
+            detections <- seq(0,1,length.out = ndetections + 1L)
+            as_relative <- TRUE
+        }
+        .check_abund_values(abund_values, x)
+        if(!.is_a_bool(as_relative)){
+            stop("'as_relative' must be TRUE or FALSE.", call. = FALSE)
+        }
+        if(as_relative && (any(detections < 0) || any(detections > 1))){
+            stop("If 'as_relative' == TRUE, detections' must be numeric ",
+                 "values between 0 and 1.", call. = FALSE)
+        }
+        #
+        x <- mia:::.agg_for_prevalence(x, rank, ...)
+        plot_data <- .get_prevalence_plot_matrix(x, abund_values, detections,
+                                                 as_relative, BPPARAM)
+        .prevalence_plotter(plot_data, 
+                            layout = "heatmap",
+                            xlab = paste0("Detection ",
+                                          ifelse(as_relative," [%]","")),
+                            ylab = ifelse(is.null(rank), "Features", rank),
+                            colour_by = "Prevalence [%]",
+                            ...)
+    }
+)
+
+.is_continuous <- function(i){
+    i <- sort(unique(i))
+    z <- round(c(i[-1L],max(i)) - i,5L)
+    length(unique(z[-length(z)])) == 1L
+}
+
+#' @importFrom BiocParallel bplapply bpisup bpstart bpstop SerialParam
+#' @importFrom SummarizedExperiment assay
+#' @importFrom tidyr pivot_longer
+#' @importFrom DelayedArray rowSums
+.get_prevalence_plot_matrix <- function(x, abund_values, detections, 
+                                        as_relative = TRUE, 
+                                        BPPARAM = BiocParallel::SerialParam()){
+    mat <- assay(x,abund_values)
+    if(as_relative){
+        mat <- mia:::.calc_rel_abund(mat)
+    }
+    FUN_prevalences <- function(d, mat){
+        rowSums(mat > d) / ncol(mat)
+    }
+    
+    if (!(bpisup(BPPARAM) || is(BPPARAM, "MulticoreParam"))) {
+        bpstart(BPPARAM)
+        on.exit(bpstop(BPPARAM), add = TRUE)
+    }
+    ans <- bplapply(detections,
+                    FUN_prevalences,
+                    mat = mat,
+                    BPPARAM = BPPARAM)
+    ans <- data.frame(ans)
+    colnames(ans) <- detections
+    lvls <- rownames(ans)[order(rowSums(ans))]
+    ans$ID <- rownames(mat)
+    ans <- ans %>%
+        pivot_longer(!.data$ID, 
+                     names_to = "detection",
+                     values_to = "prevalence")
+    colnames(ans) <- c("Y","X","colour_by")
+    ans$X <- round(as.numeric(ans$X),4) * 100
+    if(!.is_continuous(ans$X)){
+        ans$X <- factor(ans$X,
+                        unique(as.character(sort(as.numeric(ans$X)))))
+    }
+    ans$Y <- factor(ans$Y,lvls)
+    ans
+}
+
+#' @importFrom ggplot2 ggplot geom_raster geom_point geom_line
+#'   scale_fill_distiller scale_y_discrete scale_x_discrete scale_x_continuous
+#'   theme_classic
+.prevalence_plotter <- function(plot_data,
```

Comment:
Another way of visualization that is common for ecological analysis of microbiota data is the Abundance-Occupancy relationship. This is basically, plotting Abundance (x-axis) versus Prevalence (Y-axis) mostly at ASV or Genus level and coloring points by a specific taxonomic rank. E.g. figure 1AB of this article [Stopnisek et al.](https://www.biorxiv.org/content/10.1101/727461v3.full). What do you think?

Reply:
I think it would be useful. We already have examples and code in the microbiome tutorial and packages (microbiome / microbiomeutils).

Reply:
I will have a look

Reply:
See option `layout = "point"` for `plotTaxaPrevalence`. (Also in examples)

Reply:
split of the plot type into extra function